src/main/java/io/airlift/slice/Slice.java

```diff
@@ -1208,11 +1209,36 @@ public ByteBuffer toByteBuffer(int index, int length)
         checkIndexLength(index, length);
 
         if (base instanceof byte[]) {
-            return ByteBuffer.wrap((byte[]) base, (int) ((address - ARRAY_BYTE_BASE_OFFSET) + index), length);
+            int offset = (int) ((address - ARRAY_BYTE_BASE_OFFSET) + index);
+            return createHeapByteBuffer((byte[]) base, offset, length);
         }
 
+        return createDirectByteBuffer(address + index, length, reference);
+    }
+
+    private static ByteBuffer createHeapByteBuffer(byte[] base, int offset, int length)
+    {
+        try {
+            return (ByteBuffer) newHeapByteBuffer.invokeExact(base, -1, 0, length, length, offset);
```

Comment:
Rather than introducing more reflection to JDK internals, which is deprecated on JDK 9, we should use the public API `slice()`. I am doubtful that the extra object allocation would be significant in any real workload.